mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/HighLevelMqttSource.scala

```diff
@@ -0,0 +1,158 @@
+      }
+
+    createOnMaterialization[Out](mqttClientSession,
+                                 transportSettings,
+                                 restartSettings,
+                                 connectionSettings,
+                                 subscriptions,
+                                 createAckHandle)
+  }
+
+  private def createOnMaterialization[Out](
+      mqttClientSession: MqttClientSession,
+      transportSettings: MqttTransportSettings,
+      restartSettings: MqttRestartSettings,
+      connectionSettings: MqttConnectionSettings,
+      subscriptions: MqttSubscribe,
+      createAckHandle: SourceQueueWithComplete[
+        Command[Nothing]
+      ] => PartialFunction[Event[Nothing], Out]
+  ) =
+    Setup
+      .source { implicit materializer => implicit attributes =>
+        implicit val system: ActorSystem = materializer.system
+
+        constructInternals[Out](mqttClientSession,
+                                transportSettings,
+                                restartSettings,
+                                connectionSettings,
+                                subscriptions,
+                                createAckHandle)
+      }
+      .mapMaterializedValue(_.flatMap(identity)(ExecutionContexts.sameThreadExecutionContext))
+
+  private def constructInternals[Out](
+      mqttClientSession: MqttClientSession,
+      transport: MqttTransportSettings,
+      restartSettings: MqttRestartSettings,
+      connect: MqttConnectPacket,
+      subscribe: MqttSubscribe,
+      acknowledgeAndOut: SourceQueueWithComplete[Command[Nothing]] => PartialFunction[Event[Nothing], Out]
+  )(implicit system: ActorSystem, materializer: Materializer) = {
+    val mqttFlow: Flow[Command[Nothing], Either[MqttCodec.DecodeError, Event[Nothing]], NotUsed] = {
+      import restartSettings._
+      RestartFlow.onFailuresWithBackoff(minBackoff, maxBackoff, randomFactor, maxRestarts) { () =>
+        Mqtt
+          .clientSessionFlow(mqttClientSession)
+          .join(transport.connectionFlow())
+      }
+    }
+    val subscribed = Promise[immutable.Seq[(String, ControlPacketFlags)]]()
+
+    val subscribePacket = subscribe.controlPacket
+    val initCommands = immutable.Seq(
+      Command(connect.controlPacket),
+      Command(subscribePacket)
+    )
+
+    val (commands: SourceQueueWithComplete[Command[Nothing]], subscription: Source[Event[Nothing], NotUsed]) =
+      Source
+        .queue[Command[Nothing]](connect.bufferSize, OverflowStrategy.backpressure)
+        .prepend(Source(initCommands))
+        .via(mqttFlow)
+        .map {
+          case Left(decodeError) =>
+            throw new RuntimeException(decodeError.toString)
+          case Right(event @ Event(s: SubAck, _)) =>
+            val subscriptionAnswer = subscribePacket.topicFilters.map(_._1).zip(s.returnCodes)
+            subscribed.trySuccess(subscriptionAnswer)
+            event
+          case Right(event) =>
+            event
+        }
```

Comment:
In similar client code that we've written, we track `Connected`s for two purposes:

1) If we receive a `Connected` and we haven't received a `SUBACK`, our connection was likely restarted and the pending `SUBSCRIBE` was potentially dropped. Sometimes (rarely) this logic can result in two `SUBSCRIBE`s being sent to the broker, but that is benign.
2) If we receive a `Connected`, and we already have received a `SUBACK`, and `ConnAckFlags` doesn't contain `SessionPresent`, then we've reconnected to a broker that doesn't know about our session, and thus we fail the flow.